test/config-luatest/vshard_upgrade_test.lua

```diff
@@ -0,0 +1,239 @@
+    g.instance_4:start({wait_until_ready = false})
+    g.instance_5:start({wait_until_ready = false})
+    g.instance_1:wait_until_ready()
+    g.instance_2:wait_until_ready()
+    g.instance_3:wait_until_ready()
+    g.instance_4:wait_until_ready()
+    g.instance_5:wait_until_ready()
+
+    local exec = [[
+        function put(v)
+            box.space.a:insert({v.id, v.bucket_id})
+            return true
+        end
+
+        function get(id)
+            return box.space.a:get(id)
+        end
+    ]]
+    g.instance_1:eval(exec)
+    g.instance_2:eval(exec)
+    g.instance_3:eval(exec)
+    g.instance_4:eval(exec)
+end)
+
+g.after_all(function(g)
+    g.instance_1:drop()
+    g.instance_2:drop()
+    g.instance_3:drop()
+    g.instance_4:drop()
+    g.instance_5:drop()
+    treegen.clean(g)
+end)
+
+local function check_vshard(g, id_1, id_2)
+    local write_exec = string.format([[
+        vshard.router.call(1, 'write', 'put', {{id = %s, bucket_id = 1}})
+        vshard.router.call(2000, 'write', 'put', {{id = %s, bucket_id = 2000}})
+    ]], id_1, id_2)
+    g.instance_5:eval(write_exec)
+
+    local read_exec = string.format([[
+        local t = {}
+        t[1] = vshard.router.call(1, 'read', 'get', {%s})
+        t[2] = vshard.router.call(2000, 'read', 'get', {%s})
+        return t
+    ]], id_1, id_2)
+    t.helpers.retrying({timeout = 20}, function()
+        local ret = g.instance_5:eval(read_exec)
+        t.assert_equals(ret[1], {id_1, 1})
+        t.assert_equals(ret[2], {id_2, 2000})
+    end)
+end
+
+local function wait_for_names(instance_name, replicaset_name)
+    t.helpers.retrying({timeout = 20}, function()
+        local info = box.info
+        t.assert_equals(info.name, instance_name)
+        t.assert_equals(info.replicaset.name, replicaset_name)
+    end)
+end
+
+g.test_vshard_upgrade = function(g)
+    local version_exec = [[ return box.space._schema:get{'version'} ]]
+    local version = g.instance_1:eval(version_exec)
+    t.assert_equals(version, {'version', 2, 11, 1})
+    -- Assert, that tested buckets are on different shards.
+    local bucket_exec = [[ return vshard.storage.buckets_info()[%s] ]]
+    local bucket = g.instance_1:eval(string.format(bucket_exec, 2000))
+    t.assert_not_equals(bucket, nil)
+    bucket = g.instance_3:eval(string.format(bucket_exec, 1))
+    t.assert_not_equals(bucket, nil)
+
+    -- Check, that vshard works before schema upgrade.
+    check_vshard(g, 1, 2000)
+
+    -- Upgrade schema on one shard, wait for names to be applied.
+    local upgrade_exec = [[
+        box.ctl.wait_rw()
+        box.schema.upgrade()
+    ]]
+    g.instance_1:eval(upgrade_exec)
+    local rs_prefix = 'replicaset-00'
+    g.instance_1:exec(wait_for_names, {g.instance_1.alias, rs_prefix .. '1'})
+    g.instance_2:exec(wait_for_names, {g.instance_2.alias, rs_prefix .. '1'})
+    -- Check, that vshard works, when one shard has names, other - doesn't.
+    check_vshard(g, 2, 2001)
```

Comment:
We're not verifying the intermediate state: after the upgrade, but before the names apply. We discussed it f2f with @Serpentian and agreed that it is worth to do.

Reply:
It is done now.